runtime/tests/checker/purity_test.go

```diff
@@ -195,3 +196,1028 @@ func TestCheckPuritySubtyping(t *testing.T) {
+			EndPos:   ast.Position{Offset: 131, Line: 11, Column: 9},
+		})
+	})
+
+	t.Run("struct param write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		pub struct R {
+			pub(set) var x: Int
+			init(x: Int) {
+				self.x = x
+			}
+		}
+		
+		pure fun foo(_ r: R): R {
+			r.x = 3
+			return r
+		}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("struct param nested write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		pub struct R {
+			pub(set) var x: Int
+			init(x: Int) {
+				self.x = x
+			}
+		}
+		
+		pure fun foo(_ r: R): R {
+			if true {
+				while true {
+					r.x = 3
+				}
+			}
+			return r
+		}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("indeterminate write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		let a: [Int] = []
+		pure fun foo() {
+			let b: [Int] = []
+        	let c = [a, b]
+       		c[0][0] = 4
+		}
+		`)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.PurityError{}, errs[0])
+		assert.Equal(t, errs[0].(*sema.PurityError).Range, ast.Range{
+			StartPos: ast.Position{Offset: 98, Line: 6, Column: 13},
+			EndPos:   ast.Position{Offset: 104, Line: 6, Column: 19},
+		})
+	})
+
+	t.Run("indeterminate append", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		let a: [Int] = []
+		pure fun foo() {
+			let b: [Int] = []
+        	let c = [a, b]
+       		c[0].append(4)
+		}
+		`)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.PurityError{}, errs[0])
+		assert.Equal(t, errs[0].(*sema.PurityError).Range, ast.Range{
+			StartPos: ast.Position{Offset: 95, Line: 6, Column: 10},
+			EndPos:   ast.Position{Offset: 107, Line: 6, Column: 22},
+		})
+	})
+
+	t.Run("nested write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		fun foo() {
+			var a = 3
+			let b = pure fun() {
+				while true {
+					a = 4
+				}
+			}
+		}
+		`)
+
+		errs := ExpectCheckerErrors(t, err, 1)
+
+		assert.IsType(t, &sema.PurityError{}, errs[0])
+		assert.Equal(t, errs[0].(*sema.PurityError).Range, ast.Range{
+			StartPos: ast.Position{Offset: 74, Line: 6, Column: 5},
+			EndPos:   ast.Position{Offset: 78, Line: 6, Column: 9},
+		})
+	})
+
+	t.Run("nested write success", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		var a = 3
+		pure fun foo() {
+			let b = fun() {
+				a = 4
+			}
+		}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("nested scope legal write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		pure fun foo() {
+			var a = 3
+			while true {
+				a = 4
+			}
+		}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("reference write", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheck(t, `
+		struct S {
+			pub(set) var x: Int
+			init(x: Int) {
+				self.x = x
+			}
+		}
+		
+		pure fun foo(_ s: &S) {
+			s.x = 3
+		}
+		`)
+
+		errs := ExpectCheckerErrors(t, err, 1)
```

Comment:
Also, add some tests for mutating references inside containers (struct field, arrays, dictionaries).
i.e: a reference somewhere in a member-access/index-access expression chain.